File: backend/repositories/systemLogRepository.js
// backend/repositories/systemLogRepository.js
import db from "../config/db.js";

/**
 * Get system audit logs with pagination and filters
 * @param {object} filters
 * @returns {Promise<{data: Array, total: number}>}
 */
const buildTriStateWhere = (column, filterValue, queryParams) => {
  const clauses = [];
  let parsed = filterValue;
  if (typeof filterValue === "string" && filterValue.startsWith("{")) {
    try {
      parsed = JSON.parse(filterValue);
    } catch {
      /* ignore */
    }
  }

  if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
    const inc = parsed.include
      ? Array.isArray(parsed.include)
        ? parsed.include
        : Object.values(parsed.include)
      : [];
    if (inc.length > 0) {
      clauses.push(`${column} IN (?)`);
      queryParams.push(inc);
    }
    const exc = parsed.exclude
      ? Array.isArray(parsed.exclude)
        ? parsed.exclude
        : Object.values(parsed.exclude)
      : [];
    if (exc.length > 0) {
      clauses.push(`${column} NOT IN (?)`);
      queryParams.push(exc);
    }
  } else if (parsed && parsed !== "All" && parsed !== "all") {
    if (Array.isArray(parsed) && parsed.length > 0) {
      clauses.push(`${column} IN (?)`);
      queryParams.push(parsed);
    } else if (typeof parsed === "string") {
      clauses.push(`${column} = ?`);
      queryParams.push(parsed);
    }
  }
  return clauses;
};

export const getLogs = async ({
  page = 1,
  limit = 20,
  search,
  action,
  targetType,
  userId,
  startDate,
  endDate,
}) => {
  const offset = (page - 1) * limit;
  const conditions = ["1=1"];
  const params = [];

  if (search) {
    conditions.push("(target_id LIKE ? OR changes LIKE ?)");
    params.push(`%${search}%`, `%${search}%`);
  }

  const actionClauses = buildTriStateWhere("action", action, params);
  if (actionClauses.length > 0) conditions.push(...actionClauses);

  const targetClauses = buildTriStateWhere("target_type", targetType, params);
  if (targetClauses.length > 0) conditions.push(...targetClauses);

  if (userId && userId !== "all") {
    conditions.push("user_id = ?");
    params.push(userId);
  }

  if (startDate) {
    conditions.push("created_at >= ?");
    params.push(`${startDate} 00:00:00`);
  }

  if (endDate) {
    conditions.push("created_at <= ?");
    params.push(`${endDate} 23:59:59`);
  }

  const whereSql = conditions.join(" AND ");

  // Count Total
  const [countRows] = await db.query(
    `SELECT COUNT(*) as total FROM system_audit_logs WHERE ${whereSql}`,
    params,
  );
  const total = countRows[0].total;

  // Get Data
  const query = `
    SELECT l.*, u.username, u.nickname, r.name as role,
      CASE l.target_type
        WHEN 'USER' THEN (SELECT COALESCE(nickname, username) FROM users WHERE id = l.target_id)
        WHEN 'PRODUCT' THEN (SELECT name FROM products WHERE id = l.target_id)
        WHEN 'ROLE' THEN (SELECT name FROM roles WHERE id = l.target_id)
        WHEN 'LOCATION' THEN (SELECT name FROM locations WHERE id = l.target_id)
        WHEN 'CATEGORY' THEN (SELECT name FROM categories WHERE id = l.target_id)
        ELSE NULL
      END as target_name
    FROM system_audit_logs l
    LEFT JOIN users u ON l.user_id = u.id
    LEFT JOIN roles r ON u.role_id = r.id
    WHERE ${whereSql}
    ORDER BY l.created_at DESC
    LIMIT ? OFFSET ?
  `;

  const [rows] = await db.query(query, [...params, limit, offset]);

  return { data: rows, total };
};

/**
 * Create a new system audit log
 * @param {object} connection - DB Connection (Transactional)
 * @param {object} logData
 * @param {number} logData.userId
 * @param {string} logData.action - CREATE, UPDATE, DELETE, LOGIN, OTHER
 * @param {string} logData.targetType - USER, ROLE, PRODUCT, etc.
 * @param {string} logData.targetId - ID string
 * @param {object|string} logData.changes - JSON object or string description
 * @param {string} [logData.ip]
 * @param {string} [logData.userAgent]
 */
export const createLog = async (
  connection,
  { userId, action, targetType, targetId, changes, ip, userAgent },
) => {
  const changesStr =
    typeof changes === "object" ? JSON.stringify(changes) : JSON.stringify({ note: changes });

  await connection.query(
    `INSERT INTO system_audit_logs (user_id, action, target_type, target_id, changes, ip_address, user_agent, created_at)
     VALUES (?, ?, ?, ?, ?, ?, ?, NOW())`,
    [userId, action, targetType, targetId, changesStr, ip || null, userAgent || null],
  );
};

File: backend/internal/repository/system_log_repository.go
package repository

import (
	"context"
	"strings"

	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/model"
	"github.com/dps-wmhris/backend/internal/utils"
	"github.com/jmoiron/sqlx"
)

type SystemLogRepository interface {
	Create(ctx context.Context, db sqlx.ExtContext, log *model.SystemLog) error
	GetLogs(ctx context.Context, req dto.GetSystemLogsRequest) (utils.PaginatedResult[dto.SystemLogResponse], error)
}

type systemLogRepositoryImpl struct {
	db *sqlx.DB
}

func NewSystemLogRepository(db *sqlx.DB) SystemLogRepository {
	return &systemLogRepositoryImpl{db: db}
}

func (r *systemLogRepositoryImpl) Create(ctx context.Context, db sqlx.ExtContext, log *model.SystemLog) error {
	query := `
		INSERT INTO system_audit_logs (user_id, action, target_type, target_id, changes, ip_address, user_agent)
		VALUES (?, ?, ?, ?, ?, ?, ?)`
		
	res, err := db.ExecContext(ctx, query,
		log.UserID, log.Action, log.TargetType, log.TargetID, log.Changes, log.IP, log.UserAgent,
	)
	
	if err != nil {
		return err
	}
	
	id, err := res.LastInsertId()
	if err == nil {
		log.ID = int(id)
	}
	
	return err
}

func buildTriStateWhere(column string, filterValue string, conditions *[]string, params *[]interface{}) {
	if filterValue == "" || filterValue == "all" || filterValue == "All" {
		return
	}
	// For simplicity, we just handle comma separated values as IN clauses if needed
	// The node.js code handled JSON format for include/exclude. We will assume simple string match for now,
	// or IN clause if comma separated.
	if strings.Contains(filterValue, "{") {
		// Fallback for JSON strings if sent by frontend
		return
	}
	
	if strings.Contains(filterValue, ",") {
		parts := strings.Split(filterValue, ",")
		*conditions = append(*conditions, column+" IN (?)")
		*params = append(*params, parts)
	} else {
		*conditions = append(*conditions, column+" = ?")
		*params = append(*params, filterValue)
	}
}

func (r *systemLogRepositoryImpl) GetLogs(ctx context.Context, req dto.GetSystemLogsRequest) (utils.PaginatedResult[dto.SystemLogResponse], error) {
	conditions := []string{"1=1"}
	var params []interface{}

	if req.Search != "" {
		conditions = append(conditions, "(l.target_id LIKE ? OR l.changes LIKE ?)")
		searchStr := "%" + req.Search + "%"
		params = append(params, searchStr, searchStr)
	}

	buildTriStateWhere("l.action", req.Action, &conditions, &params)
	buildTriStateWhere("l.target_type", req.TargetType, &conditions, &params)

	if req.UserID != "" && req.UserID != "all" {
		conditions = append(conditions, "l.user_id = ?")
		params = append(params, req.UserID)
	}

	if req.StartDate != "" {
		conditions = append(conditions, "l.created_at >= ?")
		params = append(params, req.StartDate+" 00:00:00")
	}

	if req.EndDate != "" {
		conditions = append(conditions, "l.created_at <= ?")
		params = append(params, req.EndDate+" 23:59:59")
	}

	whereSql := strings.Join(conditions, " AND ")

	// Menggunakan generic utility untuk eksekusi query paginasi
	baseQuery := `
		SELECT l.id, l.user_id, l.action, l.target_type, l.target_id, l.changes, l.ip_address, l.user_agent, l.created_at,
			u.username, u.nickname, r.name as role,
			CASE l.target_type
				WHEN 'USER' THEN (SELECT COALESCE(nickname, username) FROM users WHERE id = l.target_id)
				WHEN 'PRODUCT' THEN (SELECT name FROM products WHERE id = l.target_id)
				WHEN 'ROLE' THEN (SELECT name FROM roles WHERE id = l.target_id)
				WHEN 'LOCATION' THEN (SELECT name FROM locations WHERE id = l.target_id)
				WHEN 'CATEGORY' THEN (SELECT name FROM categories WHERE id = l.target_id)
				ELSE NULL
			END as target_name
		FROM system_audit_logs l
		LEFT JOIN users u ON l.user_id = u.id
		LEFT JOIN roles r ON u.role_id = r.id
		WHERE ` + whereSql + `
		ORDER BY l.created_at DESC
	`

	// Tangani IN clause jika ada
	baseQuery, params, err := sqlx.In(baseQuery, params...)
	if err != nil {
		return utils.PaginatedResult[dto.SystemLogResponse]{}, err
	}
	baseQuery = r.db.Rebind(baseQuery)

	return utils.FetchPaginated[dto.SystemLogResponse](ctx, r.db, baseQuery, req.Page, req.Limit, params...)
}
